packages/runtime/container-runtime/src/gc/gcUnreferencedStateTracker.ts

```diff
@@ -45,54 +47,90 @@ export class UnreferencedStateTracker {
-				0x3b1 /* inactiveTimer still running after sweepTimer fired! */,
+				0x3b1 /* inactiveTimer still running after tombstoneTimer fired! */,
 			);
+
+			if (this.tombstoneSweepDelayMs > 0) {
+				// After the node becomes tombstone ready, start the sweep timer after which the node will be ready for sweep.
+				this.sweepTimer.restart(this.tombstoneSweepDelayMs);
+			} else {
+				this._state = UnreferencedState.SweepReady;
+			}
 		});
 
 		this.inactiveTimer = new TimerWithNoDefaultTimeout(() => {
 			this._state = UnreferencedState.Inactive;
 
-			// After the node becomes inactive, start the sweep timer after which the node will be ready for sweep.
-			if (this.sweepTimeoutMs !== undefined) {
-				this.sweepTimer.restart(this.sweepTimeoutMs - this.inactiveTimeoutMs);
+			// After the node becomes inactive, start the tombstone timer after which the node will be ready for tombstone.
+			if (this.tombstoneTimeoutMs !== undefined) {
+				this.tombstoneTimer.restart(this.tombstoneTimeoutMs - this.inactiveTimeoutMs);
 			}
 		});
+
 		this.updateTracking(currentReferenceTimestampMs);
 	}
 
 	/* Updates the unreferenced state based on the provided timestamp. */
 	public updateTracking(currentReferenceTimestampMs: number) {
 		const unreferencedDurationMs = currentReferenceTimestampMs - this.unreferencedTimestampMs;
 
-		// If the node has been unreferenced for sweep timeout amount of time, update the state to SweepReady.
-		if (this.sweepTimeoutMs !== undefined && unreferencedDurationMs >= this.sweepTimeoutMs) {
-			this._state = UnreferencedState.SweepReady;
-			this.clearTimers();
+		// Below we will set the appropriate timer (or none). Any running timers are superceded by the new currentReferenceTimestampMs
+		this.clearTimers();
+
+		// If the node has been unreferenced long enough, update the state to SweepReady.
+		if (
+			this.tombstoneTimeoutMs !== undefined &&
+			unreferencedDurationMs >= this.tombstoneTimeoutMs + this.tombstoneSweepDelayMs
+		) {
+			this._state = UnreferencedState.SweepReady; //* Try doing a bug here
+			return;
+		}
+
+		// If the node has been unreferenced long enough, update the state to TombstoneReady.
+		// Also, start a timer for the remainder of the sweep delay.
+		if (
+			this.tombstoneTimeoutMs !== undefined &&
+			unreferencedDurationMs >= this.tombstoneTimeoutMs
+		) {
+			this._state = UnreferencedState.TombstoneReady;
+
+			this.sweepTimer.restart(
+				this.tombstoneTimeoutMs + this.tombstoneSweepDelayMs - unreferencedDurationMs,
+			);
 			return;
 		}
 
-		// If the node has been unreferenced for inactive timeoutMs amount of time, update the state to inactive.
-		// Also, start a timer for the sweep timeout.
+		// If the node has been unreferenced for long enough, update the state to inactive.
+		// Also, start a timer for the remainder of the tombstone timeout.
 		if (unreferencedDurationMs >= this.inactiveTimeoutMs) {
 			this._state = UnreferencedState.Inactive;
 			this.inactiveTimer.clear();
```

Comment:
This is not needed anymore. The timers have been cleared in the beginning.

Reply:
Good catch